lib/graphql/execution/interpreter/runtime.rb

```diff
@@ -17,28 +28,81 @@ class Runtime
         # @return [GraphQL::Query::Context]
         attr_reader :context
 
-        def initialize(query:, response:)
+        def initialize(query:)
           @query = query
           @dataloader = query.multiplex.dataloader
           @schema = query.schema
           @context = query.context
           @multiplex_context = query.multiplex.context
           @interpreter_context = @context.namespace(:interpreter)
-          @response = response
-          @dead_paths = {}
-          @types_at_paths = {}
+          @result = GraphQLResultHash.new
           # A cache of { Class => { String => Schema::Field } }
           # Which assumes that MyObject.get_field("myField") will return the same field
           # during the lifetime of a query
           @fields_cache = Hash.new { |h, k| h[k] = {} }
+          # { Class => Boolean }
+          @lazy_cache = Hash.new { |h, k| h[k] = schema.lazy?(k) }
+        end
+
+        # TODO not maintain this?
+        def write_in_response(path, value)
+          response = @result
+          *keys, last = path
+          keys.each do |key|
+            if response && (response = response[key])
+              next
+            else
+              break
+            end
+          end
+          if response
+            response[last] = value
+          end
+          nil
+        end
+
+        def value_at(path)
+          response = @result
+          path.each do |key|
+            if response && (response = response[key])
+              next
+            else
+              break
+            end
+          end
+          clean_result(response)
         end
 
         def final_value
-          @response.final_value
+          clean_result(@result)
+        end
+
+        def clean_result(r)
```

Comment:
I think you could get rid of this code entirely by chopping off the response immediately when an invalid `nil` propagates during execution? You should already be at the right spot in the response object since you do set `dead` there if I understand correctly.

Maybe that would involve tracking a `parent` pointer in `GraphQLResultHash`? But that doesn't seem so bad.

Reply:
Yeah, we'd need a parent reference to do the propagation while the query is running. (Interestingly, this GraphQLResultHash is starting to look _just like_ the old `GraphQL::Query::FieldResolutionContext`! But with less overhead since it's also the result hash -- leaf fields don't create an instance of it, only objects.)

Reply:
Took a shot at it in https://github.com/rmosolgo/graphql-ruby/pull/3494/commits/96851cf73fa7ccb0dd4021d0d3f90fcec81b0d25